diff -Naur PackageKit-1.2.6/meson.build PackageKit-1.2.6-patched/meson.build
--- PackageKit-1.2.6/meson.build	2022-12-01 18:47:48.000000000 +0000
+++ PackageKit-1.2.6-patched/meson.build	2023-03-12 19:39:28.609020447 +0000
@@ -45,7 +45,7 @@
 
 elogind = []
 if get_option('elogind')
-  elogind = dependency('elogind', version: '>=229.4')
+  elogind = dependency('libelogind', version: '>=229.4')
   add_project_arguments ('-DHAVE_SYSTEMD_SD_LOGIN_H=1', language: 'c')
 endif
 
